core/src/main/scala/magnolia1/magnolia.scala

```diff
@@ -944,11 +944,46 @@ private[magnolia1] object CompileTimeState {
   }
 }
 
-object CallByNeed { def apply[A](a: => A): CallByNeed[A] = new CallByNeed(() => a) }
-final class CallByNeed[+A](private[this] var eval: () => A) extends Serializable {
+object CallByNeed {
+
+  /** Initializes a class that allows for suspending evaluation of a value until it is needed. Evaluation of a value via `.value` can only
+    * happen once. Evaluation of a value via `.valueEvaluator` will return None. For evaluating a value multiple times, please construct a
+    * CallByNeed via CallByNeed.withValueEvaluator(value)
+    */
+  def apply[A](a: => A): CallByNeed[A] = new CallByNeed(() => a, () => false)
+
+  /** Initializes a class that allows for suspending evaluation of a value until it is needed. Evaluation of a value via `.value` can only
+    * happen once. Evaluation of a value via `.valueEvaluator.map(evaluator => evaluator())` will happen every time the evaluator is called
+    */
+  def withValueEvaluator[A](a: => A): CallByNeed[A] = new CallByNeed(() => a, () => true)
+}
+
+// Both params are later nullified to reduce overhead and increase performance.
+// The supportDynamicValueEvaluation is passed as a function so that it can be nullified. Otherwise, there is no need for the function value.
+final class CallByNeed[+A] private (private[this] var eval: () => A, private var supportDynamicValueEvaluation: () => Boolean)
+    extends Serializable {
+
+  def this(eval: () => A) = this(eval, () => false)
+
+  val valueEvaluator: Option[() => A] = {
```

Comment:
shouldn't this be a `lazy val` as well? otherwise we compute this upfront when the class is constructed, not on-demand. Or maybe we can make this a `def`, which would simply call `eval()` if it's not-null, and if it's already nulled, would throw an exception saying that you can only access the value using either .value, or .valueEvaluator, but not both?

Reply:
To take a step back, this class involves the following:
1. Initial data is nullified as much as possible to keep the class lightweight.
2. Cannot change fields or class structure because of bincompat
3. The value evaluator is unused in other use cases not involving default params and is therefore adding overhead that needs to be reduced as much as possible

Ideally (something for v2), there would be two classes - something like `CallByNeedStatic` and `CallByNeedDynamic` that each extend a `CallByNeed` trait, each containing a `lazy val value` and the second containing an additional `def evaluateValue()`, as you suggested. 

Since the ideal option is not possible due to backwards compatibility requirements, I've tried to do the next best thing:
1. Add an additional param, `supportDynamicValueEvaluation`, which will ensure that the value evaluator is `Some` when the flag is set to true. The param is passed as a function so that it can later be nullified. If it is false, then the evaluator is set to `None` so that overhead is reduced.
2. I considered and ruled out making both values lazy or making `valueEvaluator` into a `def`, since both of these would involve throwing intransparent initialization errors if one `lazy val` is initialized before the other or the `lazy val` is called before the `def`. I considered it too bug prone and not user-friendly. Apis should be transparent and easy for end-users to reason about and either behavior would violate that expectation.

Reply:
Yeah I agree that two classes would be best, and that's what we'll eventually do. Re-reading the code initializing `valueEvaluator` on construction should work fine. Thanks for the explanation.